azurelinuxagent/common/cgroupapi.py

```diff
@@ -356,64 +364,28 @@ def start_extension_command(self, extension_name, command, timeout, shell, cwd,
 
         # Wait for process completion or timeout
         try:
-            try:
-                process_output = handle_process_completion(process=process,
-                                                           command=command,
-                                                           timeout=timeout,
-                                                           stdout=stdout,
-                                                           stderr=stderr,
-                                                           error_code=error_code)
-            finally:
-                with self._systemd_run_commands_lock:
-                    self._systemd_run_commands.remove(process.pid)
+            return handle_process_completion(process=process, command=command, timeout=timeout, stdout=stdout, stderr=stderr, error_code=error_code)
         except ExtensionError as e:
             # The extension didn't terminate successfully. Determine whether it was due to systemd errors or
             # extension errors.
-            systemd_failure = self._is_systemd_failure(scope, stderr)
-            process_output = read_output(stdout, stderr)
-
-            if not systemd_failure:
+            if not self._is_systemd_failure(scope, stderr):
                 # There was an extension error; it either timed out or returned a non-zero exit code. Re-raise the error
                 raise
+
+            # There was an issue with systemd-run. We need to log it and retry the extension without systemd.
+            process_output = read_output(stdout, stderr)
+            # Reset the stdout and stderr
+            stdout.truncate(0)
+            stderr.truncate(0)
+
+            if isinstance(e, ExtensionOperationError):
```

Comment:
Can this block be broken up into multiple `except` clauses? I see an inheritance tree for `ExtensionError`s; is the `self._is_systemd_failure` check on line 371 not related to the `ExtensionError` inheritance tree? i.e., can a `systemd_failure` be reported under any `ExtensionError` subtype?

Reply:
_is_systemd_failure does not look at the exception, it is unrelated to SystemdRunError

ExtensionError indicates an error on the extension, while SystemdRunError indicates an error on systemd-run; they are unrelated.